okio/jvm/src/main/java/okio/SegmentedByteString.kt

```diff
@@ -222,40 +226,65 @@ internal class SegmentedByteString(buffer: Buffer, byteCount: Int) : ByteString(
     otherOffset: Int,
     byteCount: Int
   ): Boolean {
-    var offset = offset
-    var otherOffset = otherOffset
-    var byteCount = byteCount
     if (offset < 0 || offset > size - byteCount ||
         otherOffset < 0 || otherOffset > other.size - byteCount) {
       return false
     }
     // Go segment-by-segment through this, comparing ranges of arrays.
-    var s = segment(offset)
-    while (byteCount > 0) {
-      val segmentOffset = if (s == 0) 0 else directory[s - 1]
-      val segmentSize = directory[s] - segmentOffset
-      val stepSize = minOf(byteCount, segmentOffset + segmentSize - offset)
-      val segmentPos = directory[segments.size + s]
-      val arrayOffset = offset - segmentOffset + segmentPos
-      if (!arrayRangeEquals(segments[s], arrayOffset, other, otherOffset, stepSize)) return false
-      offset += stepSize
-      otherOffset += stepSize
-      byteCount -= stepSize
-      s++
+    var otherOffset = otherOffset
+    processSegments(offset, offset + byteCount) { data, offset, byteCount ->
+      if (!arrayRangeEquals(data, offset, other, otherOffset, byteCount)) return false
+      otherOffset += byteCount
     }
     return true
   }
 
-  override fun indexOf(other: ByteArray, fromIndex: Int) = toByteString().indexOf(other, fromIndex)
-
-  override fun lastIndexOf(other: ByteArray, fromIndex: Int) = toByteString().lastIndexOf(other,
-      fromIndex)
-
   /** Returns a copy as a non-segmented byte string.  */
   private fun toByteString() = ByteString(toByteArray())
 
   override fun internalArray() = toByteArray()
 
+  /** Processes all segments, invoking `yield` with the ByteArray and range of valid data. */
+  private inline fun processSegments(
```

Comment:
Can we eliminate this overload and use default values for `beginIndex` and `endIndex` on the other?

Reply:
Certainly could, but I was concerned about the performance impact as the other function does a few extra calculations this one is able to avoid.

I'm working on a performance test which will hopefully verify that we didn't lose performance on any of these functions. I'll also test if this function is required or if the other has similar performance.